[Don't Merge] Review v0.1.15 -> v0.1.17

byeongsu-hong opened the pull request:
* general
    * bump cosmwasm to v1.5 (LTS) - stable until 2025 Apr
    * test fixes
* periphery
    * add ibcx-math package to implement pow with same logic as Osmosis using
    * add ibcx-pool package to avoid obese periphery logic and modularization
        * supports WeightedPool (GAMM) only
        * Stable Pool and CL Pool support will following
* core
    * prune portfolio assets before finalize rebalance
    * add tests

byeongsu-hong committed:
feat: periphery v2 (#56)

* wip

* deps

* trait argument

* spread factor

* add return type

* additional getter

* cfmm logic

* remove non-trait fn

* add: swap_exact_amount_* in weighted pool

* fix: revert commit

* add: swap_exact_amount_* in weighted pool

* add: calculate div Uint128 with Decimal

* fix: new pool assets assign logic & only unreachable pattern issue left

* fix: unreachable issue and unused variable

* stableswap - direct

* add: swap_exact_amount_* in weighted pool (#57)

* add: swap_exact_amount_* in weighted pool

* add: calculate div Uint128 with Decimal

* fix: new pool assets assign logic & only unreachable pattern issue left

* fix: unreachable issue and unused variable

* weighted pool

* stableswap test

* lock

* working

* lock

* right before logic test

* test weighted pool

* stable swap test & fix index issue

* fix: cal in amount given out calculation

* mint exact amount in

* fix method name

* burn exact amount out

* fix mint_exact_amount_out

* testing

* remove localnet

* cleanup

* pre mint

* wip

* warning

* publish

* v0.1.10-test

* v0.1.12

* remove faucet

* poolmanager -> gamm

* big

* fix: token_in.amount to input_amount.amount

* fix: Decimal to Decimal256

* use rev

* sort mint_msg.funds

* type casting

* sim burn exact amount in v2

* update query

* approx2

* approx3

* 0.1.14

* fix events

* bump v2

* v0.1.15-rc0

* rework simulator

* build

* target

* resolve warning

* underline

* hacky calibration

* v0.1.15

---------

Co-authored-by: steve806 <steve8060914@gmail.com>

byeongsu-hong committed:
fix!: event attribute value

byeongsu-hong committed:
fix deps

byeongsu-hong committed:
remove unwrap as many as possible

byeongsu-hong committed:
fix: codespace undefined (#58)

* inline

* downgrade std & fix

byeongsu-hong committed:
turn to workspace + bump osmosis deps

byeongsu-hong committed:
[ION-99] fix simulation error + restruct the way to testing (#59)

* parse all pools response

* simulation works

* approx pow

* remove rust_decimal

* fix zero problem

* refactor

* schema support for CL pools

* cleanup

* not to ignore unsupported pool

* pool tester

* make test-tube querier

* it works

* make test works

* fix calculation

* use non aarch wasm file

* use pre-compiled test querier

* generator

* finish build script

* support cl pool just query swap

* move to test-utils

* use upstream

* fix: apply code review

* fix: compare simulation result with actual query result

* remove pub

* move test-utils into new repo

* relative

* one file contract

* fix: review

* apply something

* use 1.69.0

* remove airdrop script

* remove pool build script

* add test data

* remove processed data

* working with test-tube

byeongsu-hong committed:
main <- develop merge (#65)

* use min_output_amount (#61)

* ignore test-querier

* 0.1.16

* feat: bump dependencies (#62)

* use "2"

* bump dependency & fix tests

* chore: clippy (#63)

* clippy

* use 0.1.4

* fix: make test pass

byeongsu-hong committed:
feat: asset pruning (#64)

* use min_output_amount (#61)

* ignore test-querier

* 0.1.16

* feat: bump dependencies (#62)

* use "2"

* bump dependency & fix tests

* chore: clippy (#63)

* clippy

* use 0.1.4

* fix: make test pass

* feat: remove asset

* bump optimizer version

* refresh locks

* 0.1.2

* tempgp

byeongsu-hong committed:
hotfix: pruning (#66)

* add proposal payload maker

* fix & add tests

hashableric commented on contracts/periphery/src/execute.rs:
```diff
@@ -1,125 +1,305 @@
     core_addr: String,
-    output_amount: Uint128,
-    input_asset: String,
+    desired_denom: String,
+    min_index_amount: Uint128,
```
This changes might effect to request payload, please check changes and notice to dongwu @byeongsu-hong

byeongsu-hong replied:
I think it's already applied on front-end side. Isn't it? @dongwu-kim

steve806 replied:
If you look below, it looks like the `mint_exact_amount_out` function still exists. 
And it's request payload is like below.
```
pub fn mint_exact_amount_out(
    deps: DepsMut,
    env: Env,
    info: MessageInfo,
    core_addr: String,
    index_amount: Uint128,
    input_denom: String,
    swap_info: Vec<SwapInfo>,
)
```
So the types of request payload was not changed.

hashableric replied:
Then, the `mint_exact_amount_in` function is newly generated? @byeongsu-hong

byeongsu-hong replied:
@hashableric yes.

hashableric commented on contracts/periphery/src/execute.rs:
```diff
@@ -1,125 +1,305 @@
+    let core_portfolio = core.get_portfolio(&deps.querier, None)?;
+
+    // input & output
+    let input_asset =
+        cw_utils::must_pay(&info, &input_denom).map(|v| coin(v.u128(), &input_denom))?;
+    let index_asset = coin(index_amount.u128(), &core_config.index_denom);
```
ask; Is there no validation for the `input_denom` is an acceptable asset?

byeongsu-hong replied:
Yep, there's no criteria to decide which denom is acceptable asset. So you can use periphery contract if you have feasible swap route for that.

hashableric commented on contracts/periphery/src/query.rs:
```diff
@@ -0,0 +1,160 @@
+    })
+}
+
+#[cfg(test)]
+mod tests {
+    use cosmwasm_std::Decimal;
+
+    #[test]
+    fn test_sad() {
+        let invert = Decimal::one()
+            .checked_div(Decimal::from_ratio(5u64, 1000u64))
+            .unwrap();
+        assert_eq!(invert, Decimal::from_ratio(200u64, 1u64));
+    }
+}
```
It seems unnecessary

byeongsu-hong replied:
just checking :)

hashableric replied:
@byeongsu-hong If you think this test is necessary, please change name from `sad` to other thing.

byeongsu-hong replied:
https://github.com/many-things/ibcx-contracts/pull/68/commits/f7fd7d5db898d7c508a7523baa19d5e927272729

hashableric commented on packages/pool/src/sim/index_in.rs:
```diff
@@ -0,0 +1,181 @@
+        };
+
+        let mut loop_count = 0;
+        loop {
+            if loop_count >= MAX_LOOP {
+                return Err(PoolError::MaxLoopExceeded);
+            }
+
+            acc_res = self.search_efficient_amount_for_output_f(
+                desired_output.clone(),
+                acc_res.max_token_out,
+                acc_res.est_min_token_in,
+            )?;
+
+            let err = Decimal::checked_from_ratio(desired_output.amount, acc_res.max_token_out)?;
+            // self.deps.api.debug(&format!(
+            //     "loop {}. input_amount={}, token_in={}, token_out={}, err={}",
+            //     loop_count,
+            //     desired_output.amount,
+            //     acc_res.est_min_token_in,
+            //     acc_res.max_token_out,
+            //     err
+            // ));
+
+            if Decimal::one() <= err
+                && desired_output.amount - acc_res.max_token_out < Uint128::from(MAX_ERROR)
+            {
+                return Ok(acc_res);
+            }
+            loop_count += 1;
+        }
+    }
```
This logic can be replaced to `for loop`. In that cases, you should choice the `for loop` for prevent unexpected infinite loop

byeongsu-hong replied:
ah okay got it

hashableric commented on packages/pool/src/sim/index_out.rs:
```diff
@@ -0,0 +1,184 @@
+                acc_res.max_token_in,
+            )?;
+
+            let err = Decimal::checked_from_ratio(desired_input.amount, acc_res.max_token_in)?;
```
It's so strange that the err variable isn't the error type.

byeongsu-hong replied:
oh it's literally mathematical error

hashableric replied:
how could `<=` operator applied for Error type? I think it is not Error type because `err` applied `<=` operator below this code @byeongsu-hong

byeongsu-hong replied:
no it's literally mathematical error

byeongsu-hong commented on .github/workflows/tester.yaml:
```diff
@@ -33,12 +33,18 @@ jobs:
-        uses: dtolnay/rust-toolchain@stable
-        with:
-          toolchain: stable
+      - name: Install Rust
+        run: rustup default 1.72.0
+
+      - name: Install Rust
+        run: rustup update nightly
+
+      - name: Install target
+        run: rustup target add wasm32-unknown-unknown
```
Noice

byeongsu-hong commented on contracts/core/src/execute/rebalance/trade/mod.rs:
```diff
@@ -97,7 +97,10 @@ mod tests {
                 Some(Rebalance::default()),
                 None,
-                Err(StdError::not_found("ibcx_core::state::config::Config").into()), // ok
+                Err(StdError::not_found(
+                    "type: ibcx_core::state::config::Config; key: [63, 6F, 6E, 66, 69, 67]",
+                )
+                .into()), // ok
```
it's just a key data of `CONFIG` storage instance. So `636F6E666967` is hex representation of `config`

hashableric replied:
Yes. Thats what I wanna jot down into comment

hashableric replied:
@byeongsu-hong Yes. Thats what I wanna jot down into comment

nicolaslara replied:
what's the reason for not sticking with the old, more readable error?

byeongsu-hong replied:
@nicolaslara it's because `not found` error comes from cw-storage-plus package. We need to wrap error to modify this error message

byeongsu-hong commented on contracts/periphery/src/error.rs:
```diff
@@ -24,6 +27,12 @@ pub enum ContractError {
     #[error("{0}")]
     CheckedFromRatioError(#[from] cosmwasm_std::CheckedFromRatioError),
 
+    #[error("{0}")]
+    IBCXMath(#[from] ibcx_math::MathError),
+
+    #[error("{0}")]
+    IBCXPool(#[from] ibcx_pool::PoolError),
```
Are you saying about naming convention or `#[error("{0}")]`?

hashableric replied:
Yup. The error formatting convension.

byeongsu-hong replied:
makes sense. applying

byeongsu-hong replied:
https://github.com/many-things/ibcx-contracts/compare/83f24735b9cd...cc415ef3f6a1

steve806 commented on contracts/periphery/src/execute.rs:
```diff
@@ -1,125 +1,305 @@
-    sim_amount_desired.sort_by(|a, b| a.denom.cmp(&b.denom));
+    let deps_ref = deps.as_ref();
+    let sim = Simulator::new(&deps_ref, &pools, &swap_info, &core_portfolio.units);
+    let sim_res = sim
+        .estimate_index_for_input(
+            desired_asset.clone(),
+            Some(min_index_amount),
+            Some(min_index_amount),
+            None,
+        )?
```
ask; Is there any reason you make 2 parameters `min_index_amount` and `init_index_amount` in `estimate_index_for_input` function? Because I couldn't find any use cases in this repo.

byeongsu-hong replied:
It's because we cannot estimate the actual swap amount in one simulation

byeongsu-hong replied:
So we need to setup initial amount to simulate it and make closer by binary search

steve806 commented on packages/math/src/pow.rs:
```diff
@@ -0,0 +1,202 @@
+    base * tmp
+}
+
+fn pow_approx(
```
👏👏👏👍

byeongsu-hong replied:
We can apply this approx logic in simulating stable swap pool :+1:

byeongsu-hong committed:
fix: reflect review of #67 (#68)

* apply review
https://github.com/many-things/ibcx-contracts/pull/67#discussion_r1464256364

* apply review
https://github.com/many-things/ibcx-contracts/pull/67#discussion_r1464324068

* error prefix for determinism

* redundant test
